File: Dynamic Programming/Hard/LET_no546_Remove Boxes.cpp
class Solution {
public:
    int memo[200][200][200] = {};
    
    int removeBoxes(vector<int>& boxes) {
        return dp(boxes, 0, boxes.size() - 1, 0);
    }
    
    int dp(vector<int>& boxes, int left, int right, int k) {
        if (left > right)
            return 0;
        if (memo[left][right][k] > 0)
            return memo[left][right][k];
        
        int i = left, d = k;
        while (i + 1 <= right && boxes[i] == boxes[i + 1]) {
            i++;
            d++;
        }
        
        int res = (d + 1) * (d + 1) + dp(boxes, i + 1, right, 0);
        for (int m = i + 1; m <= right; m++) {
            if (boxes[m] == boxes[i])
                res = max(res, dp(boxes, m, right, d + 1) + dp(boxes, i + 1, m - 1, 0));
        }
        
        return memo[left][right][k] = res;
    }
};

File: Dynamic Programming/Hard/LET_no546_Remove Boxes.py
class Solution:
    def removeBoxes(self, boxes: List[int]) -> int:
        @cache
        def dp(left, right, k):
            if left > right:
                return 0
            while left + 1 <= right and boxes[left] == boxes[left + 1]:
                left += 1
                k += 1
            res = (k + 1) * (k + 1) + dp(left + 1, right, 0)
            
            for i in range(left + 1, right + 1):
                if boxes[i] == boxes[left]:
                    res = max(res, dp(left + 1, i - 1, 0) + dp(i, right, k + 1))
            
            return res
        
        return dp(0, len(boxes) - 1, 0)